### PR TITLE
feat(sdk.v2): Support dsl.ParallelFor in v2.

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**
- Supporting `dsl.ParallelFor` in v2 sdk.
  + Caveat: Only when the loop argument is a static list (i.e. value known at compile time), can a user use `subvar` to iterate over a slice of the loop argument list using a `subvar` key.
- Fixing a bug where `deploymentConfig` could be lost.
- Fixing `LoopArguments.name_is_withitems_loop_argument()`, which has no existing usage.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/kfp/v2/compiler/compiler.py [hunk 1]
```diff
@@ -489,13 +492,132 @@ def _resolve_value_or_reference(
...
+
+      # Loop arguments are from dynamic input: pipeline param or task output
+      if _for_loop.LoopArguments.name_is_withparams_loop_argument(
+          loop_argument_name):
+
+        arg_and_var_name = (
+            _for_loop.LoopArgumentVariable
+            .parse_loop_args_name_and_this_var_name(loop_argument_name))
+        # The current IR representation is insufficient for referencing a subvar
+        # which is a key in a list of dictionaries.
```
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/compiler/compiler.py:538]: Yep, chatted offline and opened an internal bug for tracking.

File: sdk/python/kfp/v2/compiler/compiler.py [hunk 2]
```diff
@@ -489,13 +492,132 @@ def _resolve_value_or_reference(
...
+
+        assert group.items_is_pipeline_param
+        pipeline_param = group.loop_args.items_or_pipeline_param
+        input_parameter_name = pipeline_param.full_name
+
+        # Correct loop argument input type in the parent component spec.
+        # The loop argument was categorized as an artifact due to its missing
+        # or non-primitive type annotation. But it should always be String
+        # typed, as its value is a serialized JSON string.
+        dsl_component_spec.pop_input_from_component_spec(
```
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/compiler/compiler.py:552]: Also chatted offline. Currently KFP DSL doesn't support iterating over artifacts even though IR supports it. This is pending a proper design.

**Comment by numerology** [comment]
/lgtm
/approve

Thanks! @chensun

**Comment by chensun** [comment]
@numerology, another look?
I had to sync and resolve conflict -- no real conflict though, just two new methods added at the same time at the end of `sdk/python/kfp/v2/dsl/dsl_utils.py`

**Pull request merged**